.github/triage/jax_toolbox_triage/bisect.py

```diff
@@ -0,0 +1,122 @@
+            "-c",
+            f"git cat-file commit {start} && git cat-file commit {end}",
+        ],
+        policy="once_per_container",
+        workdir=dir,
+    )
+    if commits_known.returncode != 0:
+        if args.override_remotes and package in args.override_remotes:
+            remote_url = args.override_remotes.get(package, "origin")
+            fetch_cmd = ["git", "fetch", remote_url, start, end]
+            worker.check_exec(fetch_cmd, workdir=dir)
+        else:
+            # default behaviour
+            # re: https://stackoverflow.com/questions/4089430/how-to-determine-the-url-that-a-local-git-repository-was-originally-cloned-from
+            remote_url_result = worker.check_exec(
+                ["git", "config", "--get", "remote.origin.url"], workdir=dir
+            )
+            if remote_url_result.returncode == 0:
+                remote_url = remote_url_result.stdout().strip()
+                fetch_cmd = ["git", "fetch", remote_url, start, end]
+                worker.check_exec(fetch_cmd, workdir=dir)
+            else:
+                logger.error(
+                    "No remote 'origin' found and no override provided. Cannot fetch missing commits"
+                )
+                raise Exception("Cannot find commits and no remote is configured")
+
+    # detect non-linear history
+    is_ancestor_result = worker.exec(
+        ["git", "merge-base", "--is-ancestor", start, end],
+        workdir=dir,
+    )
+    is_linear = is_ancestor_result.returncode == 0
+    cherry_pick_range = {}
+
+    if not is_linear:
+        logger.info(f"Using non-linear history logic with branch {main_branch}")
+
+        # 1. find the linear range on the main branch
+        passing_and_failing_cmd = worker.check_exec(
+            [
+                "sh",
+                "-c",
+                f"git merge-base {start} {end} && git merge-base {end} {main_branch}",
+            ],
+            workdir=dir,
+        ).stdout.strip()
+        passing_main_commit, failing_main_commit = passing_and_failing_cmd.splitlines()
+
+        # 2. find commits to cherry-pick from the failing branch
+        # TODO: as an alternative approach we may need to consider `{passing_main_commit}..{start}`
+        cherry_pick_range[package] = f"{failing_main_commit}..{end}"
+
+        # 3. now we can use the main branch  commits for bisection
+        start = passing_main_commit
+        end = failing_main_commit
+
+    logger.info(
+        f"INFO: cherry_pick_range {cherry_pick_range}, start: {start} and end {end}"
+    )
+    # check if the start is the root commit. We may have to deal with the very start of the repo
+    # so we need to handle this case too
+    parent_check_result = worker.check_exec(
+        ["git", "rev-list", "--parents", "-n", "1", start], workdir=dir
+    )
+    is_root_commit = len(parent_check_result.stdout.strip().split()) == 1
+    log_range = f"{start}..{end}" if is_root_commit else f"{start}^..{end}"
```

Comment:
I doubt this is ever relevant in the real world, I would be happy to just error out in that case and avoid paying for an extra `check_exec`.